File: solutions_python/merge_two_sorted_lists/merge_two_sorted_lists.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    # Add a method to print the linked list
    def printList(self):
        current = self
        while current:
            print(current.val, end=" -> " if current.next else "\n")
            current = current.next

class Solution(object):
    def mergeTwoLists(self, list1: ListNode, list2: ListNode) -> ListNode:
        dummy: ListNode = ListNode(-1)
        current: ListNode = dummy

        while list1 is not None and list2 is not None:
            if list1.val <= list2.val:
                current.next = list1
                list1 = list1.next
            else:
                current.next = list2
                list2 = list2.next
            current = current.next

        current.next = list1 if list1 is not None else list2

        return dummy.next
    
solution = Solution()

node3L1 = ListNode(4)
node2L1 = ListNode(2, node3L1)
node1L1 = ListNode(1, node2L1)

node3L2 = ListNode(4)
node2L2 = ListNode(3, node3L2)
node1L2 = ListNode(1, node2L2)

result = solution.mergeTwoLists(node1L1, node1L2)
result.printList()

File: solutions_go/merge_two_sorted_lists/merge_two_sorted_lists.go
package main

import "fmt"

// https://leetcode.com/problems/merge-two-sorted-lists/description/
// My solution: https://leetcode.com/problems/merge-two-sorted-lists/submissions/1624112514/

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func (l *ListNode) PrintList() {
	for l != nil {
		fmt.Print(l.Val)
		if l.Next != nil {
			fmt.Print(" -> ")
		}
		l = l.Next
	}
	fmt.Print()
}

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	dummy := &ListNode{}
	current := dummy

	for list1 != nil && list2 != nil {
		if list1.Val <= list2.Val {
			current.Next = list1
			list1 = list1.Next
		} else {
			current.Next = list2
			list2 = list2.Next
		}

		current = current.Next
	}

	if list1 != nil {
		current.Next = list1
	} else {
		current.Next = list2
	}

	return dummy.Next
}

func main() {
	node3L1 := &ListNode{Val: 4}
	node2L1 := &ListNode{Val: 2, Next: node3L1}
	node1L1 := &ListNode{Val: 1, Next: node2L1}

	node3L2 := &ListNode{Val: 4}
	node2L2 := &ListNode{Val: 3, Next: node3L2}
	node1L2 := &ListNode{Val: 1, Next: node2L2}

	result := mergeTwoLists(node1L1, node1L2)
	result.PrintList()
}
